async_generator/_impl.py

```diff
@@ -263,47 +310,59 @@ def ag_frame(self):
     # Core functionality
     ################################################################
 
-    # We make these async functions and use await, rather than just regular
-    # functions that pass back awaitables, in order to get more useful
-    # tracebacks when debugging.
+    # These need to return awaitables, rather than being async functions,
+    # to match the native behavior where the firstiter hook is called
+    # immediately on asend()/etc, even if the coroutine that asend()
+    # produces isn't awaited for a bit.
+
+    def __anext__(self):
+        return self._do_it(self._it.__next__)
 
-    async def __anext__(self):
-        return await self._do_it(self._it.__next__)
+    def asend(self, value):
+        return self._do_it(self._it.send, value)
 
-    async def asend(self, value):
-        return await self._do_it(self._it.send, value)
+    def athrow(self, type, value=None, traceback=None):
+        return self._do_it(self._it.throw, type, value, traceback)
 
-    async def athrow(self, type, value=None, traceback=None):
-        return await self._do_it(self._it.throw, type, value, traceback)
+    def _do_it(self, start_fn, *args):
+        coro_state = getcoroutinestate(self._coroutine)
+        if coro_state is CORO_CREATED:
+            (firstiter, self._finalizer) = get_asyncgen_hooks()
+            if firstiter is not None:
+                firstiter(self)
 
-    async def _do_it(self, start_fn, *args):
         # On CPython 3.5.2 (but not 3.5.0), coroutines get cranky if you try
         # to iterate them after they're exhausted. Generators OTOH just raise
         # StopIteration. We want to convert the one into the other, so we need
         # to avoid iterating stopped coroutines.
         if getcoroutinestate(self._coroutine) is CORO_CLOSED:
             raise StopAsyncIteration()
-        if self.ag_running:
-            raise ValueError("async generator already executing")
-        try:
-            self.ag_running = True
-            return await ANextIter(self._it, start_fn, *args)
-        finally:
-            self.ag_running = False
+
+        async def step():
+            if self.ag_running:
+                raise ValueError("async generator already executing")
```

Comment:
What do you think of moving this check, and the `self.ag_running = True`, up out of the `async def`?

There would be one tricky thing to worry about: if a coroutine object gets `close`d or GCed before it even starts, then python doesn't enforce its normal guarantee that `finally` blocks will be executed. So if we did do this I guess we'd need to add another `yield` at the top of `step`, and immediately iterate it that far when we created the coroutine object, in order to guarantee that eventually `ag_running = False` would happen. Maybe this is more work than it's worth.

Reply:
I guess the intent would be to guard against two asend() coroutines existing simultaneously, even if one is exhausted before the second is started? Given the implementation difficulties, I don't think it's worth it, since it seems impossible to actually interleave the underlying send() calls in a way that would create confusion. (I originally put the ag_running test outside of step() with the setting-to-True still inside, and that did create potential problems, which is why I added `test_reentrance_forbidden_simultaneous_asends`.)